### ISSUE TITLE
Does not build with latest nightly rust

### ISSUE THREAD

**Issue by freiguy1** [erickt/rust-quasi#35]
I get 6 errors when building with nightly rust:

```
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:129:9: 129:97 error: mismatched types:
 expected `collections::vec::Vec<syntax::ast::TokenTree>`,
    found `core::result::Result<collections::vec::Vec<syntax::ast::TokenTree>, syntax::errors::DiagnosticBuilder<'_>>`
(expected struct `collections::vec::Vec`,
    found enum `core::result::Result`) [E0308]
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:129         parse_tts_from_source_str("<quote expansion>".to_string(), s, cx.cfg(), cx.parse_sess())
                                                                                                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:129:9: 129:97 help: run `rustc --explain E0308` to see a detailed explanation
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:139:9: 139:97 error: mismatched types:
 expected `collections::vec::Vec<syntax::ast::TokenTree>`,
    found `core::result::Result<collections::vec::Vec<syntax::ast::TokenTree>, syntax::errors::DiagnosticBuilder<'_>>`
(expected struct `collections::vec::Vec`,
    found enum `core::result::Result`) [E0308]
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:139         parse_tts_from_source_str("<quote expansion>".to_string(), s, cx.cfg(), cx.parse_sess())
                                                                                                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:139:9: 139:97 help: run `rustc --explain E0308` to see a detailed explanation
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:310:9: 314:53 error: mismatched types:
 expected `syntax::ptr::P<syntax::ast::Item>`,
    found `core::option::Option<syntax::ptr::P<syntax::ast::Item>>`
(expected struct `syntax::ptr::P`,
    found enum `core::option::Option`) [E0308]
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:310         parse::parse_item_from_source_str(
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:311             "<quote expansion>".to_string(),
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:312             s,
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:313             self.cfg(),
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:314             self.parse_sess()).expect("parse error")
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:310:9: 314:53 help: run `rustc --explain E0308` to see a detailed explanation
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:318:9: 321:83 error: mismatched types:
 expected `syntax::codemap::Spanned<syntax::ast::StmtKind>`,
    found `core::option::Option<syntax::codemap::Spanned<syntax::ast::StmtKind>>`
(expected struct `syntax::codemap::Spanned`,
    found enum `core::option::Option`) [E0308]
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:318         parse::parse_stmt_from_source_str("<quote expansion>".to_string(),
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:319                                           s,
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:320                                           self.cfg(),
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:321                                           self.parse_sess()).expect("parse error")
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:318:9: 321:83 help: run `rustc --explain E0308` to see a detailed explanation
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:325:9: 328:61 error: mismatched types:
 expected `syntax::ptr::P<syntax::ast::Expr>`,
    found `core::result::Result<syntax::ptr::P<syntax::ast::Expr>, syntax::errors::DiagnosticBuilder<'_>>`
(expected struct `syntax::ptr::P`,
    found enum `core::result::Result`) [E0308]
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:325         parse::parse_expr_from_source_str("<quote expansion>".to_string(),
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:326                                           s,
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:327                                           self.cfg(),
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:328                                           self.parse_sess())
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:325:9: 328:61 help: run `rustc --explain E0308` to see a detailed explanation
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:332:9: 335:60 error: mismatched types:
 expected `collections::vec::Vec<syntax::ast::TokenTree>`,
    found `core::result::Result<collections::vec::Vec<syntax::ast::TokenTree>, syntax::errors::DiagnosticBuilder<'_>>`
(expected struct `collections::vec::Vec`,
    found enum `core::result::Result`) [E0308]
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:332         parse::parse_tts_from_source_str("<quote expansion>".to_string(),
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:333                                          s,
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:334                                          self.cfg(),
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:335                                          self.parse_sess())
/home/freied/.multirust/toolchains/nightly/cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:332:9: 335:60 help: run `rustc --explain E0308` to see a detailed explanation
error: aborting due to 6 previous errors
Build failed, waiting for other jobs to finish...
Could not compile `quasi`.
```

**Comment by KodrAus** [comment]
That's kind of weird actually... Why was `libsyntax` changed to not use `Result`s when parsing arbitrary strings?

**Comment by ctaggart** [comment]
I think I've run into the exact same problem. I'm just trying to follow the getting started docs. https://github.com/serde-rs/serde#using-serde-with-nightly-rust-and-serde_macros

```
Camerons-iMac:atmion_json cameron$ cargo build
    Updating registry `https://github.com/rust-lang/crates.io-index`
   Compiling serde v0.7.0
   Compiling serde_codegen v0.7.0
   Compiling quasi v0.7.0
   Compiling aster v0.13.1
   Compiling num v0.1.31
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:129:9: 129:97 error: mismatched types:
 expected `collections::vec::Vec<syntax::ast::TokenTree>`,
    found `core::result::Result<collections::vec::Vec<syntax::ast::TokenTree>, syntax::errors::DiagnosticBuilder<'_>>`
(expected struct `collections::vec::Vec`,
    found enum `core::result::Result`) [E0308]
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:129         parse_tts_from_source_str("<quote expansion>".to_string(), s, cx.cfg(), cx.parse_sess())
                                                                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:129:9: 129:97 help: run `rustc --explain E0308` to see a detailed explanation
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:139:9: 139:97 error: mismatched types:
 expected `collections::vec::Vec<syntax::ast::TokenTree>`,
    found `core::result::Result<collections::vec::Vec<syntax::ast::TokenTree>, syntax::errors::DiagnosticBuilder<'_>>`
(expected struct `collections::vec::Vec`,
    found enum `core::result::Result`) [E0308]
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:139         parse_tts_from_source_str("<quote expansion>".to_string(), s, cx.cfg(), cx.parse_sess())
                                                                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:139:9: 139:97 help: run `rustc --explain E0308` to see a detailed explanation
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:310:9: 314:53 error: mismatched types:
 expected `syntax::ptr::P<syntax::ast::Item>`,
    found `core::option::Option<syntax::ptr::P<syntax::ast::Item>>`
(expected struct `syntax::ptr::P`,
    found enum `core::option::Option`) [E0308]
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:310         parse::parse_item_from_source_str(
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:311             "<quote expansion>".to_string(),
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:312             s,
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:313             self.cfg(),
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:314             self.parse_sess()).expect("parse error")
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:310:9: 314:53 help: run `rustc --explain E0308` to see a detailed explanation
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:318:9: 321:83 error: mismatched types:
 expected `syntax::codemap::Spanned<syntax::ast::StmtKind>`,
    found `core::option::Option<syntax::codemap::Spanned<syntax::ast::StmtKind>>`
(expected struct `syntax::codemap::Spanned`,
    found enum `core::option::Option`) [E0308]
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:318         parse::parse_stmt_from_source_str("<quote expansion>".to_string(),
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:319                                           s,
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:320                                           self.cfg(),
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:321                                           self.parse_sess()).expect("parse error")
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:318:9: 321:83 help: run `rustc --explain E0308` to see a detailed explanation
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:325:9: 328:61 error: mismatched types:
 expected `syntax::ptr::P<syntax::ast::Expr>`,
    found `core::result::Result<syntax::ptr::P<syntax::ast::Expr>, syntax::errors::DiagnosticBuilder<'_>>`
(expected struct `syntax::ptr::P`,
    found enum `core::result::Result`) [E0308]
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:325         parse::parse_expr_from_source_str("<quote expansion>".to_string(),
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:326                                           s,
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:327                                           self.cfg(),
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:328                                           self.parse_sess())
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:325:9: 328:61 help: run `rustc --explain E0308` to see a detailed explanation
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:332:9: 335:60 error: mismatched types:
 expected `collections::vec::Vec<syntax::ast::TokenTree>`,
    found `core::result::Result<collections::vec::Vec<syntax::ast::TokenTree>, syntax::errors::DiagnosticBuilder<'_>>`
(expected struct `collections::vec::Vec`,
    found enum `core::result::Result`) [E0308]
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:332         parse::parse_tts_from_source_str("<quote expansion>".to_string(),
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:333                                          s,
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:334                                          self.cfg(),
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:335                                          self.parse_sess())
/Users/cameron/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi-0.7.0/src/lib.rs:332:9: 335:60 help: run `rustc --explain E0308` to see a detailed explanation
error: aborting due to 6 previous errors
Build failed, waiting for other jobs to finish...
Could not compile `quasi`.

To learn more, run the command again with --verbose.
Camerons-iMac:atmion_json cameron$ rustc --version
rustc 1.9.0-nightly (c9629d61c 2016-03-10)
```

**Comment by mitsuhiko** [comment]
I locally slapped a bunch of .unwrap()s on the calls in question and that at least made it compile but I suppose that is not the long term fix.

**Comment by yberreby** [comment]
@KodrAus It's the opposite - libsyntax now uses `Result`s where it previously didn't.

See https://github.com/serde-rs/quasi/pull/36 for a possible fix.

**Comment by sjmackenzie** [comment]
`serde` depends on `quasi` thus `serde` fails to build on rust unstable. It would be marvelous if a working package could be pushed to crates.io asap :-) thank you!

**Comment by yberreby** [comment]
@sjmackenzie A new version _was_ pushed: https://crates.io/crates/quasi

You need to update your Cargo.toml to use `quasi = "0.8.0"`, though, not just Cargo.lock.

**Comment by sjmackenzie** [comment]
I just had `serde` related stuff. Nevermind, I implemented it using `rustc-serialize` :-) thanks!

**Comment by cengiz-io** [comment]
@sjmackenzie Hello.

`quasi = "0.8.0"` seems to compile with `rustc 1.9.0-nightly (02310fd31 2016-03-19)`.

```
    Updating registry `https://github.com/rust-lang/crates.io-index`
   Compiling quasi v0.8.0
   Compiling num v0.1.31
   Compiling aster v0.14.0
   Compiling serde_codegen v0.7.1
   Compiling serde v0.7.0
   Compiling quasi_codegen v0.8.0
   Compiling quasi_macros v0.8.0
   Compiling serde_json v0.7.0
   Compiling serde_macros v0.7.0
```

Before updating to `2016-03-19` I was experiencing the same problem.

**Comment by sjmackenzie** [comment]
@cengizIO thanks for letting me know! Though the code is already done and dusted: 

https://gitlab.com/sj_mackenzie/fractalide/blob/aircloak/components/example/wrangle/processchunk/convert_json_vector/Cargo.toml#L13 

It's able to tear through some 10,000 JSON files in 4.5 seconds, so it's doing fine.

**Comment by PaulGrandperrin** [comment]
Hi, it looks like compilation is broken again on the latest nightly (2016-03-25) :

```
    Updating registry `https://github.com/rust-lang/crates.io-index`
 Downloading serde_macros v0.7.0
 Downloading quasi_macros v0.8.0
   Compiling serde_codegen v0.7.1
   Compiling quasi_macros v0.8.0
/home/paulg/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi_macros-0.8.0/src/lib.rs:19:29: 19:32 error: mismatched types:
 expected `&mut syntex::Registry`,
    found `&mut rustc_plugin::registry::Registry<'_>`
(expected struct `syntex::Registry`,
    found struct `rustc_plugin::registry::Registry`) [E0308]
/home/paulg/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi_macros-0.8.0/src/lib.rs:19     quasi_codegen::register(reg);
                                                                                                                         ^~~
/home/paulg/.cargo/registry/src/github.com-88ac128001ac3a9a/quasi_macros-0.8.0/src/lib.rs:19:29: 19:32 help: run `rustc --explain E0308` to see a detailed explanation
error: aborting due to previous error
Build failed, waiting for other jobs to finish...
error: Could not compile `quasi_macros`.

To learn more, run the command again with --verbose.
```

**Comment by PaulGrandperrin** [comment]
My bad, it definitely not related to rustc-nightly (or quasi_macros probably).
I tried to bisect the bug on the nightlies from march 18, 20, 22, 24, 25 and all of them "have" the bug.
I then tried to bisect the problem with the dependencies included in my cargo.toml and I've found the (probably indirect) culprit: "rs-es":

```
[dependencies]
serde = "*"
serde_json = "*"
serde_macros = "*"
rs-es = "*" # makes compilation of quasi_macros fail
```

There must be a version conflict in the dependencies... I'll try to investigate.

**Comment by PaulGrandperrin** [comment]
The explanation is quite simple:
The build failure was caused by these two conflicting crates being compiled by dependency: quasi_macros (included by serde_macros) and quasi_codegen (included by serde_codegen itself included by rs-es).
It's true that it totally makes sense that these two conflit but the error message was less than helpful...
I don't know if it's possible but specifying at cargo's level (in cargo.toml) that these two crates conflit with each other would be great.

**Comment by cengiz-io** [comment]
@PaulGrandperrin I didn't fully understand but are you saying that two separate dependencies include the same sub dependency?

**Comment by PaulGrandperrin** [comment]
Sorry that my comment wasn't clear.
My project depends on **rs-es** and **serde_macros**.
Here is the resulting dependency graph:

```
my_project --> rs-es -> serde_codegen -> quasi_codegen
            \-> serde_macros -> quasi_macros
```

Unfortunately **quasi_macros** and **quasi_codegen** are in incompatible with each other and make the compilation of **quasi_macros** fail (cf. my first comment)

Even though it is understandable that these two crate are incompatible the issue is that the error presented to the developer is an opaque compilation failure.

A nicer way to present the problem to the developer would have been cargo complaining about these two crates being in conflit even before starting the compilation.

However I don't know if it's possible to tell cargo that two crate are in conflict.

And BTW, I resolved my immediate issue by compiling **rs-es** with a feature called nightly which use **serde_macros** instead of **serde_codegen**.

In the end I don't think there isn't anything to do on the **quasi** side, so feel free to close this issue.

**Comment by gyscos** [comment]
> Unfortunately quasi_macros and quasi_codegen are in incompatible with each other and make the compilation of quasi_macros fail (cf. my first comment)

Seeing as `quasi_macros` [depends](https://github.com/serde-rs/quasi/blob/master/quasi_macros/Cargo.toml) on `quasi_codegen`, I doubt that (maybe because of `default-features = false`?)
